hs-bindgen/src/HsBindgen/Hs/AST/ToTH.hs

```diff
@@ -0,0 +1,149 @@
+  'Storable'
+-------------------------------------------------------------------------------}
+
+instance ToTH Hs.StorableInstance where
+  type InTH Hs.StorableInstance = TH_InstanceDecl
+
+  toTH Hs.StorableInstance{
+           storableType
+         , storableSizeOf
+         , storableAlignment
+         , storablePeek
+         , storablePoke
+         } = (
+        TH.cxt []
+      , [t| Storable $(TH.conT (TH.mkName $ Hs.structName storableType)) |]
+      , sequence [
+            simpleFun 'sizeOf    [TH.wildP] $ TH.lift storableSizeOf
+          , simpleFun 'alignment [TH.wildP] $ TH.lift storableAlignment
+          , simpleFun 'peek      []         $ toTH storablePeek
+          , simpleFun 'poke      []         $ toTH storablePoke
+          ]
+      )
+
+instance ToTH Hs.PeekByteOff where
+  toTH (Hs.PeekByteOff ptr i) = [| peekByteOff ptr i |]
+
+instance ToTH Hs.PokeByteOff where
+  toTH (Hs.PokeByteOff ptr i x) = [| pokeByteOff ptr i x |]
+
+{-------------------------------------------------------------------------------
+  Structs
+-------------------------------------------------------------------------------}
+
+instance ToTH Hs.IntroStruct where
+  toTH (Hs.IntroStruct struct) = TH.conE (TH.mkName $ Hs.structConstr struct)
+
+instance DefToTH a => ToTH (Hs.ElimStruct a) where
+  toTH (Hs.ElimStruct struct x k) = do
+      fs <- mapM freshVar (Hs.structFields struct)
+      TH.caseE (TH.lift x) [
+          simpleMatch
+            (TH.conP (TH.mkName $ Hs.structConstr struct) $
+               map (TH.varP . boundVar) fs)
+            (toTH $ k fs)
+        ]
+
+{-------------------------------------------------------------------------------
+  Internal auxiliary: TH names
+-------------------------------------------------------------------------------}
+
+freshVar :: String -> Q (Const TH.Name Hs.Bound)
+freshVar x = Const <$> TH.newName x
+
+boundVar :: Const TH.Name Hs.Bound -> TH.Name
+boundVar = getConst
+
+instance TH.Lift (Const TH.Name Hs.Bound) where
+  lift (Const x) = TH.varE x
+  liftTyped = undefined
+
+{-------------------------------------------------------------------------------
+  Internal auxiliary: construct TH
+-------------------------------------------------------------------------------}
+
+type TH_InstanceDecl = (TH.CxtQ, TH.TypeQ, TH.DecsQ)
+
+instanceDecl :: TH_InstanceDecl -> TH.DecQ
+instanceDecl (cxt, typ, decs) = TH.InstanceD Nothing <$> cxt <*> typ <*> decs
+
+-- | Simple function: single clause, no guards, no where-bound variables
+simpleFun :: TH.Name -> [TH.PatQ] -> TH.ExpQ -> TH.DecQ
+simpleFun f xs b = TH.funD f [TH.clause xs (TH.normalB b) []]
+
+-- | Simple match: no guards, no where-bound variables
+simpleMatch :: TH.PatQ -> TH.ExpQ -> TH.MatchQ
+simpleMatch pat b = TH.match pat (TH.normalB b) []
+
+-- | Applicative call
+--
+-- f <$> e1 <*> .. <*> eN
+apE :: TH.ExpQ -> [TH.ExpQ] -> TH.ExpQ
+apE = \f es ->
+    case es of
+      []    -> f
+      e:es' -> go [| $f <$> $e |] es'
```

Comment:
this is `foldl'` isn't it (except you wrote` foldl`).

Reply:
Ish. I can write it as

```haskell
apE :: TH.ExpQ -> [TH.ExpQ] -> TH.ExpQ
apE = \f es ->
    case es of
      []    -> f
      e:es' -> foldl' (\x y -> [| $x (<*>) $y |]) [| $f <$> $e |] es'
```

Not sure if that is clearer, but I don't mind.